crates/testing/src/sdk.rs

```diff
@@ -0,0 +1,390 @@
+pub struct StandaloneHandle {
+    _hidden: (),
+}
+
+impl Default for StandaloneHandle {
+    fn default() -> Self {
+        let mut process = STANDALONE_PROCESS.lock().expect("STANDALONE_PROCESS Mutex is poisoned");
+        if let Some(proc) = &mut *process {
+            proc.add_user()
+                .expect("Failed to add user for running spacetime standalone process");
+        } else {
+            *process = Some(StandaloneProcess::start());
+        }
+        StandaloneHandle { _hidden: () }
+    }
+}
+
+impl Drop for StandaloneHandle {
+    fn drop(&mut self) {
+        let mut process = STANDALONE_PROCESS.lock().expect("STANDALONE_PROCESS Mutex is poisoned");
+        if let Some(proc) = &mut *process {
+            if proc
+                .sub_user()
+                .expect("Failed to remove user for running spacetime standalone process")
+            {
+                *process = None;
+            }
+        }
+    }
+}
+
+lazy_static! {
+    /// An exclusive lock which ensures we only run `spacetime generate` once for each target directory.
+    ///
+    /// Without this lock, if multiple `Test`s ran concurrently in the same process
+    /// with the same `client_project` and `generate_subdir`,
+    /// the test harness would run `spacetime generate` multiple times concurrently,
+    /// each of which would remove and re-populate the bindings directory,
+    /// potentially sweeping them out from under a compile or run process.
+    ///
+    /// This lock ensures that only one `spacetime generate` process runs at a time,
+    /// and the `HashSet` ensures that we run `spacetime generate` only once for each output directory.
+    ///
+    /// Circumstances where this will still break:
+    /// - If multiple tests want to use the same client_project/generate_subdir pair,
+    ///   but for different modules' bindings, only one module's bindings will ever be generated.
+    ///   If you need bindings for multiple different modules, put them in different subdirs.
+    /// - If multiple distinct test harness processes run concurrently,
+    ///   they will encounter the race condition described above,
+    ///   because the `BINDINGS_GENERATED` lock is not shared between harness processes.
+    ///   Running multiple test harness processes concurrently will break anyways
+    ///   because each will try to run `spacetime start` as a subprocess and will therefore
+    ///   contend over port 3000.
+    ///   Prefer constructing multiple `Test`s and `Test::run`ing them
+    ///   from within the same harness process.
+    //
+    // I (pgoldman 2023-09-11) considered, as an alternative to this lock,
+    // having `Test::run` copy the `client_project` into a fresh temporary directory.
+    // That would be more complicated, as we'd need to re-write dependencies
+    // on the client language's SpacetimeDB SDK to use a local absolute path.
+    // Doing so portably across all our SDK languages seemed infeasible.
```

Comment:
I think this is fine for now, as we continue to add tests I think we will likely run into a similar problem that we ran into with the smoketests where it starts getting extremely slow running the tests in serial rather than parallel. I think as a first pass this decision is fine

Reply:
We don't run the actual tests in serial; we only run `spacetime generate` in serial, and that we only impose on tests that use the same client project. Cargo also imposes some linearity when it comes to compiling the module, and in the Rust SDK, the client project, but again, only for tests that use the same module or the same client project.

Reply:
Ok this is fine then, thank you for the response 👍